pkg/controllers/deprovisioning/controller.go

```diff
@@ -0,0 +1,513 @@
+		var err error
+		// Capture new cluster state before each attempt
+		candidates, err := c.candidateNodes(ctx, d.shouldNotBeDeprovisioned)
+		if err != nil {
+			return ResultFailed, fmt.Errorf("determining candidate nodes, %w", err)
+		}
+		// If there are no candidate nodes, move to the next deprovisioner
+		if len(candidates) == 0 {
+			continue
+		}
+		// Sort candidates by each deprovisioner's priority, and deprovision the first one possible
+		candidates = d.sortCandidates(candidates)
+		result, err = c.executeDeprovisioning(ctx, d, candidates...)
+		if err != nil {
+			logging.FromContext(ctx).Errorf("deprovisioning nodes, %s", err)
+			continue
+		}
+		if result == ResultRetry {
+			return ResultRetry, nil
+		}
+		// Return if success, continue to other nodes otherwise.
+		if result == ResultSuccess {
+			return result, nil
+		}
+		// Continue to next deprovisioner if nothing was done
+		if result == ResultNothingToDo {
+			continue
+		}
+		return result, nil
+	}
+	// All deprovisioners did nothing, so return nothing to do
+	return ResultNothingToDo, nil
+}
+
+func (c *Controller) executeDeprovisioning(ctx context.Context, d deprovisioner, nodes ...candidateNode) (Result, error) {
+	// Given candidate nodes, compute best deprovisioning action
+	attempts := 0
+	var cmd deprovisioningCommand
+	var err error
+	success := false
+	// Each attempt will try at least one node, limit to that many attempts.
+	for attempts < len(nodes) {
+		cmd, err = d.computeCommand(ctx, attempts, nodes...)
+		if err != nil {
+			return ResultFailed, err
+		}
+		// Validate the command with a TTL
+		ok, err := c.validateCommand(ctx, cmd, d)
+		if err != nil {
+			return ResultFailed, fmt.Errorf("validating command, %w", err)
+		}
+		// If validation succeeds, execute the command. Otherwise, we need to compute another command and
+		// try the rest of the candidate nodes. If any of the remaining candidateNodes are still valid
+		// for an action, continue with deprovisioning, even though we might not be taking the most optimal decision.
+		// This is meant to stop an issue if another controller in the cluster coincidentally creates churn at the same interval.
+		if ok {
+			success = true
+			break
+		}
+		attempts++
+	}
+	// Validation for all candidateNodes failed, so nothing can be done for this deprovisioner.
+	if !success {
+		return ResultNothingToDo, nil
+	}
+	// Execute the command
+	result, err := c.executeCommand(ctx, cmd)
+	if err != nil {
+		return ResultFailed, err
+	}
+	return result, nil
+}
+
+func (c *Controller) validateCommand(ctx context.Context, cmd deprovisioningCommand, d deprovisioner) (bool, error) {
+	// Only continue with the rest of the logic if the command has an executable action for the deprovisioner
+	if !d.isExecutableCommand(cmd) {
+		return false, nil
+	}
+	// deprovisioningTTL is how long we wait before re-validating our lifecycle command
+	remainingDelay := d.getTTL() - c.clock.Since(cmd.created)
+	if remainingDelay > 0 {
```

Comment:
Ellis and I briefly talked about this logic. I think we both think there's a better way to handle this wait, but nothing to change right now. Just something to think about so that we don't hang here while we are waiting for the TTL